### PR TITLE
Target.WriteAsyncLogEvents(IList<AsyncLogEventInfo>) to public

### LINKED ISSUE #2738: Unable to extend WrapperTargetBase without converting IList<AsyncLogEventInfo> to AsyncLogEventInfo[]
Hi

I'm trying to make a target wrapper but I don't see how I'm supposed to implement the Write(IList<AsyncLogEventInfo> logEvents) method without converting the IList to an array.

We are expected to override the Write method that takes a IList (since the array one is deprecated)
`protected override void Write(IList<AsyncLogEventInfo> logEvents) {
            WrappedTarget.WriteAsyncLogEvents(logEvents);
}`

But the WriteAsyncLogEvents that takes a IList is marked as internal forcing me to convert it to an array and call the WriteAsyncLogEvents(params AsyncLogEventInfo[] logEvents) method.

Is there a better way to do this or should this be a Bug/Feature Request?

**Type**:
- Question

**NLog version**: 4.5.4

**Platform**: .Net 4.5 

In case of a BUG:

- What is the current result?
Having to convert the IList<AsyncLogEventInfo> to AsyncLogEventInfo[] to be able to call the only public method on the Wrapped Target

- What is the expected result?
To be able to call the WriteAsyncLogEvents that is currently marked as internal.

- Are there any workarounds? yes
- Is there a version in which it did work? no

### PR DESCRIPTION
**Pull request by 304NotModified**
fixes #2738

### CODE REVIEW

**Comment by 304NotModified** [comment]
@snakefoot please your opinoin.

Also not sure, maybe this is a breaking change? https://stackoverflow.com/questions/1456785/a-definitive-guide-to-api-breaking-changes-in-net

- e.g `WriteAsyncLogEvents(null)`

**Comment by snakefoot** [comment]
Yes it looks good.

File: src/NLog/Targets/Target.cs [hunk 1]
```diff
@@ -337,7 +337,7 @@ public void WriteAsyncLogEvents(params AsyncLogEventInfo[] logEvents)
         /// Writes the array of log events.
         /// </summary>
         /// <param name="logEvents">The log events.</param>
-        internal void WriteAsyncLogEvents(IList<AsyncLogEventInfo> logEvents)
+        public void WriteAsyncLogEvents(IList<AsyncLogEventInfo> logEvents)
```
Comment (NLog-SonarQube) [inline src/NLog/Targets/Target.cs]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 19 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)

**Comment by NLog-SonarQube** [comment]
SonarQube analysis reported 5 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 3 major

Watch the comments in this conversation to review them.

#### 4 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [Target.cs#L194](https://github.com/NLog/NLog/blob/66b850745e40c98ecd5fde735ec6475a53e0e533/src/NLog/Targets/Target.cs#L194): Refactor this method to reduce its Cognitive Complexity from 36 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS3776)
1. ![MAJOR][MAJOR] [Target.cs#L198](https://github.com/NLog/NLog/blob/66b850745e40c98ecd5fde735ec6475a53e0e533/src/NLog/Targets/Target.cs#L198): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1066)
1. ![MAJOR][MAJOR] [Target.cs#L497](https://github.com/NLog/NLog/blob/66b850745e40c98ecd5fde735ec6475a53e0e533/src/NLog/Targets/Target.cs#L497): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1066)
1. ![MAJOR][MAJOR] [Target.cs#L738](https://github.com/NLog/NLog/blob/66b850745e40c98ecd5fde735ec6475a53e0e533/src/NLog/Targets/Target.cs#L738): Merge this if statement with the enclosing one. [rule](https://sonarcloud.io/coding_rules#rule_key=csharpsquid%3AS1066)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Comment by snakefoot** [comment]
How does it become a breaking change?

**Comment by snakefoot** [comment]
Well I think it is a good thing to punish those who pass null, or other random stuff.

**Comment by 304NotModified** [comment]
> Well I think it is a good thing to punish those who pass null, or other random stuff.

maybe, but not in a patch release. 

Testing this change currently. If it fails, I will move this to 4.5.6 with a new protected method name

**Comment by 304NotModified** [comment]
testing it. At least it's only a source breaking change and not a binary.

**Comment by 304NotModified** [comment]
tested it with:

- `AsyncLogEventInfo[]` -> params
- `null` -> params
- `List<AsyncLogEventInfo>` -> ilist

**Pull request merged**